src/types.ts

```diff
@@ -1,7 +1,32 @@
 export type UnbindFn = () => void;
 
-export type Binding = {
-  type: string;
-  listener: EventListenerOrEventListenerObject;
+type AnyFunction = (...args: any[]) => any;
+
+type GetEventType<Target extends EventTarget, Type extends string> = Target extends unknown
+  ? `on${Type}` extends keyof Target
+    ? GetEventTypeFromListener<
+        // remove types that aren't assignable to `AnyFunction`
+        // so that we don't end up with union like `MouseEvent | Event`
+        Extract<Target[`on${Type}`], AnyFunction>
+      >
+    : Event
+  : never;
+
+type GetEventTypeFromListener<T extends AnyFunction> = T extends (this: any, event: infer U) => any
```

Comment:
technically an event listener can be a function or an object with a `handleEvent` property (my current types don't account for this either